python/python/bystro/covariance/hypothesis_hd.py

```diff
@@ -0,0 +1,873 @@
+            df += 1
+            statistic += result["statistic"] ** 2
+            if result["c"] == 1:
+                reject += 1
+
+    if df < 10:
+        pvalue = 0
+        statistic = 10000
+    else:
+        pvalue = 1 - chi2.cdf(statistic, df)
+
+    return {
+        "decision": int(reject > rng.binomial(k, alpha)),
+        "statistic": statistic,
+        "pvalue": pvalue,
+        "df": df,
+        "reject": reject,
+    }
+
+
+def calibration(
+    n1: int,
+    n2: int,
+    p: int,
+    n: int,
+    alpha: float = 0.05,
+    const: float = 0.5,
+    iterations: int = 100,
+    K: int = 100,
+    seed: int = 2021,
+) -> float:
+    """
+    Perform calibration to find threshold.
+
+    Parameters
+    ----------
+    n1 : int
+        Size of the first sample.
+    n2 : int
+        Size of the second sample.
+    p : int
+        Number of columns.
+    n : int
+        Sample size.
+    alpha : float, optional
+        Significance level, by default 0.05.
+    const : float, optional
+        Constant value, by default 0.5.
+    iterations : int, optional
+        Number of iterations, by default 100.
+    K : int, optional
+        Number of sub-iterations, by default 100.
+    seed : int, optional
+        Random seed, by default 2021.
+
+    Returns
+    -------
+    threshold : float
+        Threshold value.
+    """
+    rng = np.random.default_rng(seed)
+    values: List[float] = []
+    for _ in range(iterations):
+        X = rng.normal(0, 1, (n1, p))
+        Y = rng.normal(0, 1, (n2, p))
+        for _ in range(K):
+            value = two_sample_test_(
+                X,
+                Y,
+                n,
+                alpha=alpha,
+                const=const,
+                mode="calib",
+                seed=rng.integers(int(1e6)),
+            )
+            if isinstance(value, float):
+                values.append(value)
+    return float(np.quantile(values, 1 - alpha))
+
+
+def c_tuning(
+    X: np.ndarray,
+    Y: np.ndarray,
+    n: int,
+    thres: Optional[float] = None,
+    alpha: float = 0.05,
+    epsilon: float = 0.05,
+    K: int = 500,
+    seed: int = 2021,
+) -> dict:
+    """
+    Tune the constant C.
+
+    Parameters
+    ----------
+    X : np.ndarray
+        First matrix.
+    Y : np.ndarray
+        Second matrix.
+    n : int
+        Sample size.
+    thres : Optional[float], optional
+        Threshold value, by default None.
+    alpha : float, optional
+        Significance level, by default 0.05.
+    epsilon : float, optional
+        Tolerance value, by default 0.05.
+    K : int, optional
+        Number of iterations, by default 500.
+    seed : int, optional
+        Random seed, by default 2021.
+
+    Returns
+    -------
+    tuning_result : dict
+        Tuned constant C and rates.
+    """
+    rng = np.random.default_rng(seed)
+    Cs = np.arange(1, 31) / 10
+    rates = []
+
+    if thres is None:
+        thres = 2.6
```

Comment:
Why do we set this default? Could we logger.warn("Threshold `thres` not specified, setting %d", 2.6)?